icefoganalytics/internal-data-portal
# Fix Scope Chaining Usage In Controllers

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/101

Relates to:
- https://github.com/sequelize/sequelize/issues/17337

# Context

**Is your feature request related to a problem? Please describe.**
This code pattern does not work:
```ts
const where = this.query.where as WhereOptions<AccessRequest>
const filters = this.query.filters as Record<string, unknown>

const scopedAccessRequests = AccessRequestsPolicy.applyScope(AccessRequest, this.currentUser)

let filteredAccessRequests = scopedAccessRequests
if (!isEmpty(filters)) {
  Object.entries(filters).forEach(([key, value]) => {
    filteredAccessRequests = filteredAccessRequests.scope({ method: [key, value] })
  })
}

const totalCount = await filteredAccessRequests.count({ where })
const accessRequests = await filteredAccessRequests.findAll({
  where,
  limit: this.pagination.limit,
  offset: this.pagination.offset,
})
```

I thought you could chain `.scope` calls, but you cannot.
The current code completely removes policy scoping when filters are in effect.
Even if the scopes chained, the policy scope should be applied last to give it highest precedence.

**Describe the solution you'd like**
I want to be able to merge filter scopes and policy scopes in a safe way.

I've come up with a pattern that works
```ts
const where = this.query.where as WhereOptions<AccessRequest>
const filters = this.query.filters as Record<string, unknown>

const scopes: BaseScopeOptions[] = []
if (!isEmpty(filters)) {
  Object.entries(filters).forEach(([key, value]) => {
    scopes.push({ method: [key, value] })
  })
}
const scopedAccessRequests = AccessRequestsPolicy.applyScope(scopes, this.currentUser)

const totalCount = await scopedAccessRequests.count({ where })
const accessRequests = await scopedAccessRequests.findAll({
  where,
  limit: this.pagination.limit,
  offset: this.pagination.offset,
})
```

Given
```ts
// See api/node_modules/sequelize/types/model.d.ts -> Model -> scope
export type BaseScopeOptions = string | ScopeOptions

export const POLICY_SCOPE_NAME = "policyScope"

export function PolicyFactory<M extends Model>(modelClass: ModelStatic<M>) {
  const policyClass = class Policy extends BasePolicy<M> {
    static applyScope(scopes: BaseScopeOptions[], user: User): ModelStatic<M> {
      this.ensurePolicyScope()

      return modelClass.scope([...scopes, { method: [POLICY_SCOPE_NAME, user] }])
    }

    /**
     * Just in time scope creation for model class.
     * TODO: to have scope creation occur at definition time, instead of execution time.
     */
    static ensurePolicyScope() {
      if (Object.prototype.hasOwnProperty.call(modelClass.options.scopes, POLICY_SCOPE_NAME)) {
        return
      }

      modelClass.addScope(POLICY_SCOPE_NAME, this.policyScope.bind(modelClass))
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static policyScope(user: User): FindOptions<Attributes<M>> {
      throw new Error("Derived classes must implement policyScope method")
    }
  }

  return policyClass
}
```

**Describe alternatives you've considered**
Switching to Sequelize 7 does not fix this.
Switching to Rails would fix this, but I'm not allowed, and it would be a lot more work. 

**Additional context**
This issue is in effect in many of the repos I've been working on. I should definitely be writing more controller tests.
Why didn't I write more controller tests? Because I don't have a good pattern or framework for writing them yet, so they are very hard to write. Also Jest is very slow, switching to Vitest will make writing tests easier.

Effects these files:
- api/src/controllers/access-requests-controller.ts
- api/src/controllers/dataset-entries-controller.ts
- api/src/controllers/datasets-controller.ts
- api/src/controllers/users-controller.ts
- api/src/controllers/users/search-controller.ts

# Implementation

1. Add just in time policy scope creation from policy scope class method of policy.
    i.e. `Model.addScope('policyScope', Policy.policyScope.bind(Model))`
    I'd prefer to add the scope at load time, instead of run time, but can't figure out how to make that work.
2. Set up policy types will to complain unless you set:
    ```ts
    static policyScope(user: User): FindOptions<Attributes<SomeSpecificModel>> {}
    ```
    I'd prefer if the types were generated magically, but can't get it to work.
    At least this way they still fail correctly.
3. Found that I can't use scopes inside of other scopes, and decided that duplicating and inlining the code was the least painful way to fix it.
    The alternatives where:
    a. Add a method to the class that returns the scope, and use it in both places.
       Best if the scope is used in many places.
    b. `AccessRequest.scope({ method: ["withDatasetOwnerId", user.id] })._scope`.
       this is a hack loading info from a private method, and is not recommended.
       But it _would_ work, so good to be aware of it.
    c. Increase the complexity of the scope generator to handle returning a scope.
       This vastly increases the complexity of the code, and types, so probably not a good choice
       long term.

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Navigate around, check that nothing is broken.
